waiter/integration/waiter/authentication_test.clj

```diff
@@ -126,3 +128,152 @@
+            set-cookie (str (get headers "set-cookie"))
+            assertion-message (str {:headers headers
+                                    :set-cookie set-cookie
+                                    :target-url target-url})]
+        (assert-response-status response 200)
+        (is (= (retrieve-username) (str body)))
+        (let [{:strs [x-waiter-auth-method]} headers]
+          (is (not= "jwt" x-waiter-auth-method) assertion-message)
+          (is (not (string/blank? x-waiter-auth-method)) assertion-message))
+        (is (= (retrieve-username) (get headers "x-waiter-auth-user")) assertion-message)
+        (assert-auth-cookie set-cookie assertion-message))
+      (log/info "JWT authentication is disabled"))))
+
+(defn- create-token-name
+  [waiter-url service-id-prefix]
+  (str service-id-prefix "." (subs waiter-url 0 (string/index-of waiter-url ":"))))
+
+(deftest ^:parallel ^:integration-fast test-jwt-authentication-token-realm
+  (testing-using-waiter-url
+    (if (jwt-auth-enabled? waiter-url)
+      (let [waiter-host (-> waiter-url sanitize-waiter-url utils/authority->host)
+            host (create-token-name waiter-url (rand-name))
+            service-parameters (assoc (kitchen-params) :name (rand-name))
+            token-response (post-token waiter-url (assoc service-parameters
+                                                    :run-as-user (retrieve-username)
+                                                    "token" host))
+            _ (assert-response-status token-response 200)
+            access-token (retrieve-access-token host)
+            request-headers {"authorization" (str "Bearer " access-token)
+                             "host" host
+                             "x-forwarded-proto" "https"}
+            {:keys [port]} (waiter-settings waiter-url)
+            target-url (str waiter-host ":" port)
+            {:keys [headers service-id] :as response}
+            (make-request-with-debug-info
+              request-headers
+              #(make-request target-url "/status" :disable-auth true :headers % :method :get))
+            set-cookie (str (get headers "set-cookie"))
+            assertion-message (str {:headers headers
+                                    :service-id service-id
+                                    :set-cookie set-cookie
+                                    :target-url target-url})]
+        (try
+          (with-service-cleanup
+            service-id
+            (assert-response-status response 200)
+            (is (= "jwt" (get headers "x-waiter-auth-method")) assertion-message)
+            (is (= (retrieve-username) (get headers "x-waiter-auth-user")) assertion-message)
+            (assert-auth-cookie set-cookie assertion-message))
+          (finally
+            (delete-token-and-assert waiter-url host))))
+      (log/info "JWT authentication is disabled"))))
+
+(deftest ^:parallel ^:integration-fast test-bypass-jwt-authentication-token-realm
```

Comment:
It would be great if we could have some integration tests that result in failures, both unauthorized and forbidden.

Reply:
The integration tests show that our JWT behavior is correct all the way to the client, which is valuable.  My first thought was that we didn't test this because it was "hard", even though we probably both agree it would be useful.

Reply:
Added integration tests for both unauthorized and forbidden responses.